src/erlcloud_ec2.erl

```diff
@@ -3574,31 +3583,65 @@ ec2_query(Config, Action, Params, ApiVersion) ->
                                   Config#aws_config.ec2_port,
                                   "/", QParams, "ec2", Config).
 
-default_config() -> erlcloud_aws:default_config().
+% Exported Query Function with parameter handling
+% Query takes in: 
+% - an aws_config
+%   - an ec2 action string: list of possible are found here https://docs.aws.amazon.com/AWSEC2/latest/APIReference/OperationList-query-ec2.html
+% - a map of query parameters: An example can be found here https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeImages.html
+    % NOTE: Any fields in the documentation that end in ".N", may just store a list - No need to include ".N".   
+
+-spec query(aws_config(), string(), map(), query_opts()) -> ok_error().
+query(Config, Action, Params, Opts) ->
+    % NewOpts = maps:merge(?OPT_DEFAULTS, Params),
+    ApiVersion= maps:get(version, Opts, ?NEW_API_VERSION),
+    Filter = maps:get(filter, Opts, []),
+    ResponseFormat = maps:get(response_format, Opts, undef),
+    erlcloud_aws:parse_response(do_query(Config, Action, Params, Filter, ApiVersion), ResponseFormat).
+
+do_query(Config, Action, MapParams, Filter, ApiVersion) -> 
+    Params = prepare_action_params(MapParams, Filter),
+    case ec2_query(Config, Action, Params, ApiVersion) of
+        {ok, Results} ->
+            {ok, Results};
+        {error, {http_error, 412, _, _}} -> 
+            {ok, dry_run_success};
```

Comment:
Which API returns a 412? This behavior is definitely worth documenting.

Reply:
All AWS API's return 412 when a dry run was successful - maybe we should be more explicit and check whether we expect a dry run response?

Reply:
Ah okay, I had looked at https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html and didn't see an exact 4XX series error code, just text that a 4XX series error code is returned and that the description would be in the response body.